cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll

```diff
@@ -0,0 +1,590 @@
+ */
+private predicate boundFlowStepSsa(
+  NonPhiOperand op2, Operand op1, int delta, boolean upper, Reason reason
+) {
+  op2.getDefinitionInstruction().getAnOperand().(CopySourceOperand) = op1 and
+  (upper = true or upper = false) and
+  reason = TNoReason() and
+  delta = 0
+  or
+  exists(IRGuardCondition guard |
+    guard = boundFlowCond(op2, op1, delta, upper, _) and
+    reason = TCondReason(guard)
+  )
+}
+
+/**
+ * Gets a condition that tests whether `op` is bounded by `bound + delta`.
+ *
+ * If the condition evaluates to `testIsTrue`:
+ * - `upper = true`  : `op <= bound + delta`
+ * - `upper = false` : `op >= bound + delta`
+ */
+private IRGuardCondition boundFlowCond(NonPhiOperand op, NonPhiOperand bound, int delta, boolean upper,
+  boolean testIsTrue)
+{
+  exists(Operand compared |
+    result.comparesLt(compared, bound, delta, upper, testIsTrue) and
+    result.controls(op.getInstruction().getBlock(), testIsTrue) and
+    valueNumber(compared.getDefinitionInstruction()) = valueNumber(op.getDefinitionInstruction())
+  )
+  // TODO: strengthening through modulus library
+}
+
+/**
+ * Gets a condition that tests whether `op` is bounded by `bound + delta`.
+ * 
+ * - `upper = true`  : `op <= bound + delta`
+ * - `upper = false` : `op >= bound + delta`
+ */
+private IRGuardCondition boundFlowCondPhi(PhiOperand op, NonPhiOperand bound, int delta, boolean upper,
+  boolean testIsTrue)
+{
+  exists(Operand compared |
+    result.comparesLt(compared, bound, delta, upper, testIsTrue) and
+    result.controlsEdge(op.getPredecessorBlock().getLastInstruction(), op.getInstruction().getBlock(), testIsTrue) and
+    valueNumber(compared.getDefinitionInstruction()) = valueNumber (op.getDefinitionInstruction())
+  )
+  or
+  exists(Operand compared |
+    result.comparesLt(compared, bound, delta, upper, testIsTrue) and
+    result.controls(op.getPredecessorBlock(), testIsTrue) and
+    valueNumber(compared.getDefinitionInstruction()) = valueNumber (op.getDefinitionInstruction())
+  )
+  // TODO: strengthening through modulus library
+}
+
+private newtype TReason =
+  TNoReason() or
+  TCondReason(IRGuardCondition guard) { possibleReason(guard) }
+
+/**
+ * A reason for an inferred bound. This can either be `CondReason` if the bound
+ * is due to a specific condition, or `NoReason` if the bound is inferred
+ * without going through a bounding condition.
+ */
+abstract class Reason extends TReason {
+  abstract string toString();
+}
+class NoReason extends Reason, TNoReason {
+  override string toString() { result = "NoReason" }
+}
+class CondReason extends Reason, TCondReason {
+  IRGuardCondition getCond() { this = TCondReason(result) }
+  override string toString() { result = getCond().toString() }
+}
+
+/**
+ * Holds if a cast from `fromtyp` to `totyp` can be ignored for the purpose of
+ * range analysis.
+ */
+private predicate safeCast(IntegralType fromtyp, IntegralType totyp) {
+  fromtyp.getSize() <= totyp.getSize() and
```

Comment:
Shouldn't this be strict in the case of casting from an unsigned type to a signed type?  E.g. I wouldn't imagine casting from `unsigned int` to `int` would qualify as a safe cast.